### PR TITLE
Integrating Bog Analytics into AAU with Loggers

### LINKED ISSUE #321: [Bog Analytics] - Integrating Bog Analytics into AAU
## Overview

We are going to track events in Angels among us as a trial run for our new Bog Analytics tool : )

**Be sure to create a new branch before working on your ticket!!**

## Acceptance Criteria

**Setup:**
- [x] Watch this video to set up the repository + understand how we are going to set up bog analytics into AAU
https://youtu.be/SmE9JrMNkgE
- [x] Setup a Logger file where you export a development, staging, and production logger
    - [x] Authenticate the loggers with the Client API key

**Events:**
- [x] Click Event: Number of clicks on a post - this will help us track how much attention a dog is getting
    - [x] object id: `dog_{dog name}`
    - [x] user id: randomly generated string is fine 
- [x] Click Event: Number of clicks on each filter - this will help us see which filters are being used and which ones are not
    - [x] object id: `filter_{filter name}`
    - [x] user id: randomly generated string is fine 
- [x] Visit Event: Tracking the user agent of visits to [fosters.angelsrescue.org](http://fosters.angelsrescue.org/)
    - [x] See: https://developer.mozilla.org/en-US/docs/Web/API/Window/navigator
    - [x] the `useEffect` hook will be valuable on this page because you can log events on initial load with it: https://react.dev/
reference/react/useEffect 
    - [x] page url: `/`
    - [x] user id: The user agent you log

## Notes
- You need 3 different loggers - one for development (local environments), one for staging (preview environments), and one for production (real life environments :0)
- use the URL to determine which logger to use. For example:
```
if (window.location.href.includes("localhost")){
    developmentLogger.logClickEvent({...})
} else if (window.location.href.includes("angels-among-us.netlify.app")){
    stagingLogger.logClickEvent({...})
} else if (window.location.href.includes("fosters.angelsrescue.org")){
    productionLogger.logClickEvent({...})
}
```
- To initialize and authenticate the Analtyics logger (you'll obviously need to import the necessary classes/enums):
```
// development logger initialization example
AnalyticsLogger developmentLogger = new AnalyticsLogger({environment: EventEnvironment.DEVELOPMENT});
developmentLogger.authenticate(process.env.PUBLIC_NEXT_BOG_ANALYTICS_CLIENT_API_KEY);
```
- Bog Analytics NPM Page: https://www.npmjs.com/package/bog-analytics

### PR DESCRIPTION
**Pull request by wonhochoi1**
Integrating Bog Analytics into AAU with Loggers

Issue Number(s): #321 

What does this PR change and why?

### Checklist

- [ ] Requirements have been implemented
- [ ] Acceptance criteria are met
- [ ] Database schema docs have been updated or are not necessary
- [ ] Code follows design and style guidelines
- [ ] Commits follow guidelines (concise, squashed, etc)
- [x] Relevant reviewers (Senior Dev/EM/Designers) have been assigned to this PR

### Critical Changes

- Database change/migration to run
- Environment config change
- Breaking API change

### Testing

Enumerate steps to test the functionality of your ticket. This should include edge cases and testing of error handling, if applicable.

### CODE REVIEW

**Review by SamratSahoo — APPROVED** [review]
lgtm but I am asking katsuki to review too

File: web/utils/analytics-logger.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+const productionLogger = new AnalyticsLogger({ environment: EventEnvironment.PRODUCTION });
+
+const apiKey = process.env.NEXT_PUBLIC_BOG_ANALYTICS_CLIENT_API_KEY as string;
+developmentLogger.authenticate(apiKey);
+stagingLogger.authenticate(apiKey);
+productionLogger.authenticate(apiKey);
+
+export function getAnalyticsLogger() {
+    return window.location.href.includes("https://fosters.angelsrescue.org") ? productionLogger
+        : window.location.href.includes("angels-among-us.netlify.app") ? stagingLogger
```
Comment (katsukixyz) [inline web/utils/analytics-logger.ts]: Check for `"staging"` instead - `"angels-among-us.netlify.app"` host is used by PR deployment previews as well, which is a development environment

Comment (katsukixyz) [reply web/utils/analytics-logger.ts]: This project also has an env variable on Netlify builds `NEXT_PUBLIC_CONTEXT` that provides the environment at runtime

File: web/pages/post/[postId].tsx [hunk 2]
```diff
@@ -84,6 +87,24 @@ function PostPage({
...
+        objectId: `dog_${postData?.name}`,
+        userId: (Math.random() + 1).toString(36).substring(7),
+      });
+
+    }
+
+    if (postData?.name) {
+      logDogName().then().catch();
+    }
+  }, [JSON.stringify(postData)])
```
Comment (katsukixyz) [inline web/pages/post/[postId].tsx]: Change dep array to just `postData`

File: web/pages/post/[postId].tsx [hunk 3]
```diff
@@ -84,6 +87,24 @@ function PostPage({
...
+      // Log the click event
+      logger.logClickEvent({
+        objectId: `dog_${postData?.name}`,
+        userId: (Math.random() + 1).toString(36).substring(7),
+      });
+
+    }
+
+    if (postData?.name) {
+      logDogName().then().catch();
```
Comment (katsukixyz) [inline web/pages/post/[postId].tsx]: Consider just calling the method without the then or catch

File: web/components/Users/Search.tsx [hunk 4]
```diff
@@ -106,7 +117,8 @@ export default function Search({
                         label: val,
                       })
                     )}
-                    onChange={(newVals) =>
+                    onChange={async (newVals) => {
+                      await logFilterClickEvent('Foster Type');
```
Comment (katsukixyz) [inline web/components/Users/Search.tsx]: Nit: Consider removing the `await` here and just call the logging function as normal since the code after it doesn't require it to proceed

File: web/components/Users/Search.tsx [hunk 5]
```diff
@@ -147,7 +159,8 @@ export default function Search({
                         label: val,
                       })
                     )}
-                    onChange={(newVals) =>
+                    onChange={async (newVals) => {
+                      await logFilterClickEvent('Medical');
```
Comment (katsukixyz) [inline web/components/Users/Search.tsx]: See above

File: web/components/Users/Search.tsx [hunk 6]
```diff
@@ -185,7 +198,8 @@ export default function Search({
                       value: key,
                       label: val,
                     }))}
-                    onChange={(newVals) =>
+                    onChange={async (newVals) => {
+                      await logFilterClickEvent('Breed');
```
Comment (katsukixyz) [inline web/components/Users/Search.tsx]: Nit above

File: web/components/Users/Search.tsx [hunk 7]
```diff
@@ -216,7 +230,8 @@ export default function Search({
                       value: key,
                       label: val,
                     }))}
-                    onChange={(newVals) =>
+                    onChange={async (newVals) => {
+                      await logFilterClickEvent('Age');
```
Comment (katsukixyz) [inline web/components/Users/Search.tsx]: oops there were more of these than i thought

File: web/components/Users/Search.tsx [hunk 8]
```diff
@@ -247,7 +262,8 @@ export default function Search({
                       value: key,
                       label: val,
                     }))}
-                    onChange={(newVals) =>
+                    onChange={async (newVals) => {
+                      await logFilterClickEvent('Size');
```
Comment (katsukixyz) [inline web/components/Users/Search.tsx]: here

File: web/components/Users/Search.tsx [hunk 9]
```diff
@@ -278,7 +294,8 @@ export default function Search({
                       value: key,
                       label: val,
                     }))}
-                    onChange={(newVals) =>
+                    onChange={async (newVals) => {
+                      await logFilterClickEvent('Gender');
```
Comment (katsukixyz) [inline web/components/Users/Search.tsx]: here

File: web/components/Users/Search.tsx [hunk 10]
```diff
@@ -315,7 +332,8 @@ export default function Search({
                     options={Object.entries(goodWithLabels).map(
                       ([key, val]) => ({ value: key, label: val })
                     )}
-                    onChange={(newVals) =>
+                    onChange={async (newVals) => {
+                      await logFilterClickEvent('Goodwith')
```
Comment (katsukixyz) [inline web/components/Users/Search.tsx]: holy moly

File: web/components/Users/Search.tsx [hunk 11]
```diff
@@ -345,7 +363,8 @@ export default function Search({
                     options={Object.entries(behavioralLabels).map(
                       ([key, val]) => ({ value: key, label: val })
                     )}
-                    onChange={(newVals) =>
+                    onChange={async (newVals) => {
+                      await logFilterClickEvent('Behavioral')
```
Comment (katsukixyz) [inline web/components/Users/Search.tsx]: here

File: web/components/Users/Search.tsx [hunk 12]
```diff
@@ -53,6 +54,16 @@ type SearchProps = {
   setSearched: Dispatch<SetStateAction<boolean>>;
 };
 
+const logFilterClickEvent = async (filterName: string) => {
```
Comment (katsukixyz) [inline web/components/Users/Search.tsx]: At a high level I think that logging clicks on this particular screen is a bit redundant since we can track the backend queries to see what the user was searching for, but it's fine since this is more of a PoC and a trial run more than anything

File: web/pages/_app.tsx [hunk 13]
```diff
@@ -111,6 +113,21 @@ const theme = extendTheme({
...
+    const logUserAgent = async () => {
+      const logger = getAnalyticsLogger();
+
+      logger.logVisitEvent({
+        userId: getBrowserName(navigator.userAgent),
+        pageUrl: '/'
+      });
+    }
+
+    logUserAgent().then().catch()
```
Comment (katsukixyz) [inline web/pages/_app.tsx]: Mentioned earlier

File: web/pages/post/[postId].tsx (hunk 2 above)
Comment (SamratSahoo) [reply web/pages/post/[postId].tsx]: ngl my editor was giving me a weird error when I was doing this (but now its not for whatever reason) so this was a quick hack lol

File: web/components/Users/Search.tsx (hunk 12 above)
Comment (SamratSahoo) [reply web/components/Users/Search.tsx]: I bet you don't get cool graphs tho 😈
